### ISSUE TITLE
Attempting to change tracks in multi-angle adaptive HLS playlist

### ISSUE THREAD

**Issue by benwilber** [google/ExoPlayer#2600]
I am trying to change the selected video track in a multi-angle adaptive HLS playlist as described in [Listing 13 here](https://developer.apple.com/library/content/technotes/tn2288/_index.html#//apple_ref/doc/uid/DTS40012238-CH1-ALTERNATE_MEDIA) "Variant playlist with 3 different camera angles and different bit rates."

I have created a test stream with 3 different "angles" each having 3 adaptive variants [here](http://ll-streams.s3.amazonaws.com/multi-stream.m3u8):

```
#EXTM3U

#EXT-X-MEDIA:TYPE=VIDEO,GROUP-ID="3000k",NAME="Red",AUTOSELECT=YES,DEFAULT=YES
#EXT-X-MEDIA:TYPE=VIDEO,GROUP-ID="3000k",NAME="Green",AUTOSELECT=YES,DEFAULT=NO,URI="green/index-f1-v1-a1.m3u8"
#EXT-X-MEDIA:TYPE=VIDEO,GROUP-ID="3000k",NAME="Blue",AUTOSELECT=YES,DEFAULT=NO,URI="blue/index-f1-v1-a1.m3u8"

#EXT-X-MEDIA:TYPE=VIDEO,GROUP-ID="5000k",NAME="Red",AUTOSELECT=YES,DEFAULT=YES
#EXT-X-MEDIA:TYPE=VIDEO,GROUP-ID="5000k",NAME="Green",AUTOSELECT=YES,DEFAULT=NO,URI="green/index-f2-v1-a1.m3u8"
#EXT-X-MEDIA:TYPE=VIDEO,GROUP-ID="5000k",NAME="Blue",AUTOSELECT=YES,DEFAULT=NO,URI="blue/index-f2-v1-a1.m3u8"

#EXT-X-MEDIA:TYPE=VIDEO,GROUP-ID="8000k",NAME="Red",AUTOSELECT=YES,DEFAULT=YES
#EXT-X-MEDIA:TYPE=VIDEO,GROUP-ID="8000k",NAME="Green",AUTOSELECT=YES,DEFAULT=NO,URI="green/index-f3-v1-a1.m3u8"
#EXT-X-MEDIA:TYPE=VIDEO,GROUP-ID="8000k",NAME="Blue",AUTOSELECT=YES,DEFAULT=NO,URI="blue/index-f3-v1-a1.m3u8"


#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=613606,RESOLUTION=1920x1080,CODECS="avc1.640028,mp4a.40.2",VIDEO="3000k"
red/index-f1-v1-a1.m3u8

#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=815785,RESOLUTION=1920x1080,CODECS="avc1.640028,mp4a.40.2",VIDEO="5000k"
red/index-f2-v1-a1.m3u8

#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=1110022,RESOLUTION=2560x1440,CODECS="avc1.640032,mp4a.40.2",VIDEO="8000k"
red/index-f3-v1-a1.m3u8
```

Unfortunately I'm not able to determine from the `TrackSelector` API how I can actually change to a different "angle" as a result of user button press, for instance.  Is this possible with ExoPlayer v2.3.0 presently?

Thanks!

**Comment by sergiomartinez4** [comment]
is there a way to achieve that in ExoPlayer r2.5.2?

**Comment by sapanda** [comment]
Bumping up to see if this is possible yet in ExoPlayer 2.8.2. Thanks!

**Comment by AquilesCanta** [comment]
Still on the radar but not on the roadmap yet, unfortunately. Thanks for your patience!

**Comment by jaganPeloton** [comment]
Hello,

I have a similar use-case where I am defining alternative renditions (HLS spec https://tools.ietf.org/html/draft-pantos-http-live-streaming-20#section-4.3.4.2.1).

How can I use ExoPlayer on Android to select audio streams from a group - This is mainly to support multi-language audio over HLS streaming?

Thanks.

**Comment by AquilesCanta** [comment]
Multi-language HLS should just work. Try it on our demo app and let us know if you run into any issues.

**Comment by WassimErriha** [comment]
I'm working on a very similar feature as the OP where I'm trying to make a seamless track selection to a different camera-angle.
I have tried two approaches so far:
**1.** I have tried to implement a ManualTrackSelector by customizing `AdaptiveTrackSelector`, and passing a message to the playback thread to make a seamless switch. This solution is described in #2250. The switch is seamless as expected, but the time between the selection and the switch can be quite long(maybe LoadControl needs to be configured differently ?). Also, `ExoPlayer.ExoPlayerComponent()` and ExoPlayer.sendMessages() are now deprecated ...
**2.** The second approach was to use two players. The idea was that the helper player loads the media  that was selected and seeks to the current position, then makes the switch only after it reaches `STATE_READY`. The seeking in this approach didn't work as expected.
With the decoder reuse in the upcoming release (Exoplayer 2.10), i might just revert to using a single instance since the latency is now down to 110ms (props to the team for this !)
Will the next release facilitate multi-angle track selection or renditions ? 
If not, is there a way to customize Exoplayer to support such use case in the meantime ?
Thank you!

**Comment by AquilesCanta** [comment]
Hi @WassimErriha, this issue tracks support for multiple video renditions in HLS (which is currently being worked on as per commit references above). 

Achieving seamless switching between video renditions is a different issue, and I suggest you open a new issue for that.

**Comment by WassimErriha** [comment]
Bumping up for supporte in version 2.11
Thank you!

**Comment by tdc-w** [comment]
@WassimErriha 
I'm also trying to realize seamless switching so that buffering does not occur when switching camera angle.

>I have tried to implement a ManualTrackSelector by customizing AdaptiveTrackSelector, and passing a message to the playback thread to make a seamless switch. 

I would like to try this implementation. Would you please share the specific sample code?

**Comment by icbaker** [comment]
Closing as a duplicate of https://github.com/androidx/media/issues/3216 to move this to the media3 tracker.
